hnn_core/optimization.py

```diff
@@ -0,0 +1,427 @@
+    params : dict
+        The params dictionary.
+    opt_dpls : dict
+        Dictionary with keys exp_dpl and best for
+        the experimental dipole and best dipole.
+
+    Returns
+    -------
+    avg_rmse: float
+        Weighted RMSE between data in dpl and exp_dpl
+    """
+    print("Optimization step %d, iteration %d" % (opt_params['cur_step'] + 1,
+                                                  opt_params['optiter'] + 1))
+
+    from .parallel_backends import _BACKEND, JoblibBackend
+    if _BACKEND is None:
+        _BACKEND = JoblibBackend(n_jobs=1)
+
+    # set parameters
+    for param_name, test_value in zip(opt_params['ranges'].keys(), new_params):
+        params[param_name] = test_value
+
+    # run the simulation, but stop early if possible
+    params['tstop'] = opt_params['opt_end']
+    net = Network(params, add_drives_from_params=True)
+    dpls = _BACKEND.simulate(net, n_trials=1)
+    # avg_dpl = average_dipoles(dpls)
+    avg_dpl = dpls[0].copy()
+    avg_rmse = rmse(avg_dpl, opt_dpls['exp_dpl'],
+                    tstart=opt_params['opt_start'],
+                    tstop=opt_params['opt_end'],
+                    weights=opt_params['weights'])
+
+    if avg_rmse < opt_params['stepminopterr']:
+        best = "[best] "
+        opt_params['stepminopterr'] = avg_rmse
+        opt_dpls['best_dpl'] = avg_dpl
+    else:
+        best = ""
+
+    print("%sweighted RMSE: %.2f over range [%3.3f-%3.3f] ms" %
+          (best, avg_rmse, opt_params['opt_start'], opt_params['opt_end']))
+
+    opt_params['optiter'] += 1
+
+    return avg_rmse  # nlopt expects error
+
+
+def run_optimization(maxiter, param_ranges, optrun):
+
+    cons = list()
+    x0 = list()
+    for idx, param_name in enumerate(param_ranges):
+        x0.append(param_ranges[param_name]['initial'])
+        cons.append(
+            lambda x: x[idx] <= param_ranges[param_name]['maxval'])
+        cons.append(
+            lambda x: x[idx] >= param_ranges[param_name]['minval'])
+
+    result = fmin_cobyla(optrun, cons=cons, rhoend=1e-4,
+                         x0=x0, maxfun=maxiter)
+    return result
+
+
+def optimize_evoked(params, exp_dpl, maxiter,
+                    timing_range_multiplier=3.0, sigma_range_multiplier=50.0,
+                    synweight_range_multiplier=500.0, decay_multiplier=1.6):
+    """Optimize drives to generate evoked response.
+
+    Parameters
+    ----------
+    params : dict
+        The initial params
+    exp_dpl : instance of Dipole
+        The target experimental dipole.
+    maxiter : int
+        The maximum number of of simulations to run for optimizing
+        one "chunk".
+    timing_range_multiplier : float
+        The scale of timing values to sweep over.
+    sigma_range_multiplier : float
+        The scale of sigma values to sweep over.
+    synweight_range_multiplier : float
+        The scale of input synaptic weights to sweep over.
+    decay_multiplier : float
+        The decay multiplier.
```

Comment:
What is the `decay_multiplier` referring to here?

Reply:
It's part of the calculation for the time windows to optimize over. I set it empirically so it should be an optional parameter if the user knows better.